scd/usbtoken.c

```diff
@@ -0,0 +1,737 @@
+ * the file called "COPYING".
+ *
+ * Contact Information:
+ * Fraunhofer AISEC <trustme@aisec.fraunhofer.de>
+ */
+#include "usbtoken.h"
+#include "ssl_util.h"
+
+#include "sc-hsm-cardservice.h"
+#include <ctapi.h>
+
+// #define LOGF_LOG_MIN_PRIO LOGF_PRIO_TRACE
+#include "common/macro.h"
+#include "common/mem.h"
+#include "common/file.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+#define TOKEN_MAX_AUTH_CODE_LEN 16
+#define TOKEN_KEY_LEN 32 /* must be coordinated with ssl_util.c */
+
+#define MAX_CT_READERS_SIZE 4096
+
+/* TODO: investigate enforcement done by hardware token */
+#define USBTOKEN_MAX_WRONG_UNLOCK_ATTEMPTS 3
+
+static unsigned short g_ctn = 0;
+
+static unsigned char requesticc[] = { 0x20, 0x12, 0x00, 0x01, 0x00 };
+
+static unsigned char skd_dskkey[] = { 0xA8, 0x2F, 0x30, 0x13, 0x0C, 0x11, 0x44, 0x69, 0x73, 0x6B,
+				      0x45, 0x6E, 0x63, 0x72, 0x79, 0x70, 0x74, 0x69, 0x6F, 0x6E,
+				      0x4B, 0x65, 0x79, 0x30, 0x08, 0x04, 0x01, 0x01, 0x03, 0x03,
+				      0x07, 0xC0, 0x10, 0xA0, 0x06, 0x30, 0x04, 0x02, 0x02, 0x00,
+				      0x80, 0xA1, 0x06, 0x30, 0x04, 0x30, 0x02, 0x04, 0x00 };
+
+static unsigned char algo_dskkey[] = { 0x91, 0x01, 0x99 };
+
+struct usbtoken {
+	char *serial;
+
+	unsigned short ctn;  // card terminal number
+	unsigned short port; // usb port of token reader
+
+	bool locked;			// whether the token is locked or not
+	unsigned wrong_unlock_attempts; // wrong consecutive password attempts
+
+	// the authentication code is cached as long as the token remains unlocked
+	unsigned char *auth_code;
+	size_t auth_code_len;
+};
+
+/**
+ * Derive an authentication code from a parining secret and a user pin/passwd.
+ * TODO: use an actual KDF
+ * @param
+ *
+ * @return the length of the authentication code in Byte
+ */
+static int
+get_auth_code(const char *pin, const unsigned char *pair_sec, size_t pair_sec_len,
+	      unsigned char *auth_code_buf, size_t buf_len)
+{
+	int ofs = 0;
+
+	int pin_len = strlen(pin);
+
+	ASSERT(auth_code_buf);
+
+	TRACE("USBTOKEN: pairing_sec_len: %d", (int)pair_sec_len);
+	TRACE("USBTOKEN: pin_len: %d", (int)pin_len);
+
+	if ((pin == NULL) || (strlen(pin) == 0)) {
+		ERROR("No PIN provided");
+		return -1;
+	}
+
+	if ((pin_len + pair_sec_len) > buf_len) {
+		ERROR("PIN and pairing secret combined must not exceed %d",
+		      TOKEN_MAX_AUTH_CODE_LEN);
+		return -1;
+	}
+
+	if (pair_sec != NULL) {
+		memcpy(auth_code_buf, pair_sec, pair_sec_len);
+		ofs = pair_sec_len;
+	}
+
+	memcpy(auth_code_buf + ofs, (unsigned char *)pin, pin_len);
+	ofs += pin_len;
+
+	return ofs;
+}
+
+/*
+ * Request card
+ */
+static int
+requestICC(int ctn)
+{
+	TRACE("USBTOKEN: requestICC");
+
+	unsigned char Brsp[260];
+	unsigned short lr;
+	unsigned char dad, sad;
+	int rc;
+
+	dad = 1; /* Reader */
+	sad = 2; /* Host */
+	lr = sizeof(Brsp);
+
+	rc = CT_data((unsigned short)ctn, &dad, &sad, sizeof(requesticc),
```

Comment:
this is because of inconsistency in the code from [1], where different types for ctn (int vs unsigned short) are used in the cardservice.c and ctapi.c.

[1] https://github.com/CardContact/sc-hsm-embedded